### ISSUE TITLE
Is it possible to associate a repo with a a profile?

### ISSUE THREAD

**Issue by justinmchase** [gabe565/gh-profile#3]
It would be really nice if just issuing commands relative to a particular repo always used a particular profile.

So I could have two terminals opened up and just `cd` to two different repos and based on the configurations, each one used a different profile.

**Comment by yermulnik** [comment]
I'm not affiliated with this repo owner, I'm just a user like you, though I think this won't be possible until is supported by GH CLI as such.
Though you might want to look into tools like `direnv` or `envrc` to manage configurations on a per-dir basis, so that when you chdir the tool triggers shell actions for you (which though isn't really "two terminals opened up and just cd to two different repos and based on the configurations, each one used a different profile").

**Comment by yermulnik** [comment]
> but haven't thought of an elegant solution yet.

The only one I can think of is e.g. `PROMPT_COMMAND` in Bash or alikes in other shells.
Though there are solutions out there already that adhere to "one tool that does job well" principle. As in tackling profile auto-switch may be a very time and effort consiming part of the project while there are other tools that do this job right. Especially given that neither GIT CLI, not GH CLI does have such a functionality yet...

**Comment by gabe565** [comment]
After some thought, this should be doable using [direnv](https://direnv.net) and `GH_CONFIG_DIR`. Each profile is created in `$HOME/.config/gh/profiles/` so I could add a symlink from each profile to the global config and it should work. I can add a subcommand to generate the `.envrc` if I'm able to get it working.

**Comment by yermulnik** [comment]
I reckon you're leaning into overcomplicating the tool with what should be handled by other means. What's requested here needs not just symlinks (while executing `gh profile switch <profile>` via `direnv`'s `.envrc` is a way better than tackling symlinks since there are other ways to switch auth creds), but also needs triggering this kind of "automation" not just on every command line submission (given chdir is already handled by `direnv`, so "automation" on every cmdline submission may be implemented via `PROMPT_COMMAND`) but during GH CLI invocation as well (which is not implemented by GH CLI at the moment and hence this and alike solutions to switch creds).
Assume you switched profile in `/path/to/dir1`, though want to use another creeds profile in `/path/to/another/dir` just like that and you issue e.g. `gh issue list` in this other dir — this would use creds that you switched to in the beginning which is not want requester wants. So you either need to run some arbitrary command or submit empty command line to, shall we say, trigger `gh profile switch <other_profile>` ebmedded into `PROMPT_COMMAND` or need to wait for GH to implement this into GH CLI 🤷🏻 Everything else is going to be poorly reliable quirks and workarounds 😢
So the easiest and fastest approach for requester would be a tiny shell script to trigger `gh profile switch ...` based on current working dir called via `PROMPT_COMMAND` (with the drawback of a need to issue e.g. empty command line to trigger script execution embedded into `PROMPT_COMMAND`).

> Each profile is created in `$HOME/.config`

I already mentioned this in the other issues: you might better rely on `$XDG_CONFIG_HOME` instead of `$HOME/.config` for compatibility and portability. Though it's up to you to decide =)

**Comment by gabe565** [comment]
> I reckon you're leaning into overcomplicating the tool with what should be handled by other means.

Definitely not. I'm a fan of single-responsibility so this change will rely on direnv to handle per-shell overrides.

> What's requested here needs not just symlinks (while executing gh profile switch <profile> via direnv's .envrc is a way better than tackling symlinks since there are other ways to switch auth creds), but also needs triggering this kind of "automation" not just on every command line submission (given chdir is already handled by direnv, so "automation" on every cmdline submission may be implemented via PROMPT_COMMAND) but during GH CLI invocation as well (which is not implemented by GH CLI at the moment and hence this and alike solutions to switch creds).
Assume you switched profile in /path/to/dir1, though want to use another creeds profile in /path/to/another/dir just like that and you issue e.g. gh issue list in this other dir — this would use creds that you switched to in the beginning which is not want requester wants. So you either need to run some arbitrary command or submit empty command line to, shall we say, trigger gh profile switch <other_profile> ebmedded into PROMPT_COMMAND or need to wait for GH to implement this into GH CLI 🤷🏻 Everything else is going to be poorly reliable quirks and workarounds 😢

That is not how this will be implemented. This change _will not_ do any changes to the filesystem. That would cause unexpected behavior since `gh profile switch` modifies the global config. As you mentioned, if a user opened a git repo, it would swap the profile for other open shells, which is undesirable. It would also slow down a user's prompt since filesystem access is slow.

Instead, this change will be implemented so that direnv sets the `$GH_CONFIG_DIR` flag on a per-repo basis. Then, opening a specific git repo would override its profile for the gh cli, while other shells would keep the global config. No filesystem changes are needed and other open shells would still use the globally activated gh profile.

> So the easiest and fastest approach for requester would be a tiny shell script to trigger gh profile switch ... based on current working dir called via PROMPT_COMMAND (with the drawback of a need to issue e.g. empty command line to trigger script execution embedded into PROMPT_COMMAND as PROMPT_COMMAND gets exectued only on cmdline submission).

I would not suggest using this for the same reasons I mentioned above.

> I already mentioned this in the other issue: you might better rely on $XDG_CONFIG_HOME instead of directly on $HOME/.config for compatibility and portability. Though it's up to you to decide =)

That has already been completed in d185690edb673c8e2418cff2773ab5e6210658a1. As I mentioned in my response to your comment, the gh cli checks for `$GH_CONFIG_DIR`, then falls back to `$XDG_CONFIG_HOME/gh`, then falls back to `$HOME/.config/gh`. gh-profile behaves the same way.

**Comment by yermulnik** [comment]
> this change will rely on direnv to handle per-shell overrides.

Oh, I have misunderstood you previously. Sorry for that.

> Instead, this change will be implemented so that direnv sets the `$GH_CONFIG_DIR` flag on a per-repo basis.

Yay, that sounds great indeed 👍🏻 
Would you mind considering `gh-profile`-specific var instead of a parent tool's sort of "global" env var? So that `gh-profile` only tackles what it owns. I think it should be a better choice in a longterm to avoid any kind of conflict, confusion or clash. E.g. `GH_PROFILE_CONFIG_DIR` might be a good candidate. And consequently `GH_PROFILE_CONFIG_DIR` would be equal to `$HOME/.config/gh/profiles` (inc all that stuff around `GH_CONFIG_DIR` and `XDG_CONFIG_HOME`).

_on a per-repo basis_ — did you mean "per-dir" or you're thinking of something else? Please apology my curiosity =)

> As I mentioned in my response to your comment

Oops, I might have missed that comment. My bad. Apologies for the noise 🤷🏻

**Comment by gabe565** [comment]
Don't be sorry! I love getting feedback so keep it coming 🙂

> I'm assuming direnv here is used as a "common name" for any tool of this kind which helps to manage environment (so that users are not locked to a particular tool of someone else's choice).

That's a good point. I'll probably add a flag to write the necessary env to stdout and another flag to write it to `.envrc`. That way, it'll be simple to set up with direnv but any other env solution would work.

> Would you mind considering gh-profile-specific var instead of a parent tool's sort of "global" env var? So that gh-profile only tackles what it owns. I think it should be a better choice in a longterm to avoid any kind of conflict, confusion or clash. E.g. GH_PROFILE_CONFIG_DIR might be a good candidate. And consequently GH_PROFILE_CONFIG_DIR would be equal to $HOME/.config/gh/profiles (inc all that stuff around GH_CONFIG_DIR and XDG_CONFIG_HOME).

Hmm maybe. Since the tool needs to be able to change the gh config either way, it would add some complexity. I won't add a separate config for now but will keep it in mind if there's a solid reason for it in the future.

> on a per-repo basis — did you mean "per-dir" or you're thinking of something else? Please apology my curiosity =)

You're right, I did mean "per-dir". I guess I was imagining that a specific git repo _probably_ doesn't use multiple GitHub accounts, but either way would be doable.

> Oops, I might have missed that comment. My bad. Apologies for the noise 🤷🏻

Not a problem! The comment I was referencing is https://github.com/gabe565/gh-profile/commit/7637db42bb001e2008836fd3d8b1b2417b665d33#r86842386. It can get hard to keep track when multiple conversations are happening at once 😅

**Comment by gabe565** [comment]
This has been implemented in [v1.2.0](https://github.com/gabe565/gh-profile/releases/tag/v1.2.0)!

[Direnv](https://direnv.net) or another directory-based env tool is required. Once installed, can run `gh profile switch --local-dir NAME` to place a local profile override in `.envrc`.

This has worked very well during my testing, but I'd really appreciate some feedback!

Only bug I've come across is if you run `gh profile switch --local-dir` in the same directory multiple times, new overrides will be added to the end of `.envrc` instead of the existing one being replaced. This still works fine since the last line takes precedence, but it's still ugly. I considered replacing the existing line but decided to keep it simple for the initial release.

**Comment by justinmchase** [comment]
Um... awesome! You guys are not messing around. I'll play with it tomorrow.

**Comment by yermulnik** [comment]
hey @gabe565 
> `f, err := os.OpenFile(".envrc", os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)`

I'm not quite sure how to do this in Golang, though it might be a good idea to consult user's current `umask` value (if exists) and fall-back to `0644` if `umask` isn't customized.
Yeah, I know it's quire rare to customize it, though I'm that one who does set it to allow owner access only 🤷🏻 

> I considered replacing the existing line but decided to keep it simple for the initial release.

Yep, multiplying same string isn't nice =) To not make it too complex you might take a look into putting some "anchors" around it in next release, something like this:
```bash
# --> START: gh profile
# Auto-generated by `gh profile switch --local-dir <NAME>`
export …
# <-- STOP: gh profile
```

Overall indeed cool stuff! Thanks! Will give it a try definitely.

By the way, is it possible to not expand user home path but use `$HOME` instead so that it has `export GH_CONFIG_DIR="$HOME/.config/gh/profiles/<NAME>"`? So that when someone copies the file or entire dir to another workstation, there's no need to adjust `.envrc` file?
It's — again — quite rare use case, though every so often I need to use my personal laptop to do some work stuff, for what I just `rsync` entire work dir hierarchy between these two, though my username differs on personal laptop from the one I've got on my work laptop.
Thanks!

**Comment by gabe565** [comment]
> I'm not quite sure how to do this in Golang, though it might be a good idea to consult user's current `umask` value (if exists) and fall-back to `0644` if `umask` isn't customized.
> Yeah, I know it's quire rare to customize it, though I'm that one who does set it to allow owner access only 🤷🏻

Go should respect your current umask! I'm not familiar enough with umasks enough to test myself. I believe Go creates the file with permissions like `perm & ~umask` so maybe I should create the file with `0666` and let the system adjust for the umask accordingly? That seems bad from a permissions standpoint, though so I'm not sure.

> Yep, multiplying same string isn't nice =) To not make it too complex you might take a look into putting some "anchors" around it in next release, something like this:

That could be useful! Although since it's a single line, I think it will be pretty simple to find `export GH_CONFIG_DIR=` and replace the value. I do like that it adds the command that generated the line.

> By the way, is it possible to not expand user home path but use `$HOME` instead so that it has `export GH_CONFIG_DIR="$HOME/.config/gh/profiles/<NAME>"`? So that when someone copies the file or entire dir to another workstation, there's no need to adjust `.envrc` file?

Yeah I like that idea. I will implement it, but I'll also account for `$XDG_CONFIG_HOME`.

**Comment by yermulnik** [comment]
> Go should respect your current umask!

Just checked and yep it does! Cool:
```bash
> ll
total 16384
drwx------+ 2 yz yz  4096 Oct 20 21:32 ./
drwxr-xr-x  7 yz yz 12288 Oct 19 22:01 ../

> gh profile switch --local-dir personal
🔧 Activating local dir profile: personal
direnv: loading /tmp/yz/qqq/.envrc
direnv: export +GH_CONFIG_DIR

> ll .envrc
-rw------- 1 yz yz 61 Oct 20 21:32 .envrc

> gh profile switch --local-dir personal
🔧 Activating local dir profile: personal
ℹℹ️️  Profile already active: personal

> cat .envrc
export GH_CONFIG_DIR='/home/yz/.config/gh/profiles/personal'
```

btw, no multiplication of record 🥳 

> Yeah I like that idea. I will implement it, but I'll also account for `$XDG_CONFIG_HOME`.

Marvelous! Thanks Gabe!

**Comment by gabe565** [comment]
@yermulnik I've implemented your suggestions in [v1.2.1](https://github.com/gabe565/gh-profile/releases/tag/v1.2.1). Still looking for feedback on local dir configs so let me know what you think. Thanks!

**Comment by yermulnik** [comment]
@gabe565 Looks great. I'll report back issues encountered if any 😺
My only concern is (given I haven't took a look into the local configs code) if GH CLI (or any extension like yours) adds something else into the `GH_CONFIG_DIR` (like additional config files and/or directories), then with the current approach `gh-profile` would need to take these into consideration and hence duplication of files/dirs to allow local profile work via `GH_CONFIG_DIR` change. So... maybe what you were looking into re the symlinks starts to make sense to me 🤔 With another concern about symlinks being not available in e.g. Windows. Though this seems to be quite another story.
Let's see how this works and adjust on the go if there are any changes/updates from GH CLI or another GH CLI extension into the `GH_CONFIG_DIR` hierarchy.

I'd say this issue got its resolution implemented in full and just perfectly! Thanks a lot Gabe!

**Comment by gabe565** [comment]
> Looks great. I'll report back issues encountered if any 😺

Thank you!

> if GH CLI [...] adds something else into the `GH_CONFIG_DIR` [...], then with the current approach `gh-profile` would need to take these into consideration

I was concerned about that that, too. For the moment, the gh cli respects the config dir for everything. If they introduce a new config, different profiles _should_ have their own version of that config by default.

Either way, I have an update monitor enabled for the gh cli so I'll keep an eye on updates and will open an issue if I ever see a breaking change.

> With another concern about symlinks being not available in e.g. Windows.

I'd like to test this extension out on Windows. Go's symlink API says it supports Windows and apparently Windows has supported symlinks since Vista (which is news to me), so I believe it would work, but it's completely untested at the moment. 😂

> I'd say this issue got its resolution implemented in full and just perfectly! Thanks a lot Gabe!

That's what I was hoping for! It's been working well on my end for the past few days so I think this is safe to close. Thank you.

**Comment by yermulnik** [comment]
> If they introduce a new config, different profiles should have their own version of that config by default.

Yep, this is what I meant with "need to take these into consideration and hence duplication of files/dirs". Like some other GH CLI extension would have had an arbitrary dir in place like `gh-profile` did so it needs to be duplicated in scope of local profile dir option of `gh-profile` =) Don't wanna imagine that dir to have some large volume of files, like a clone of a repo with too many files of 1Gb each 🤣 That's a theory though any theory may come true 🤣 Who the hell know what people may do by an accident =))))

> Either way, I have an update monitor enabled for the gh cli so I'll keep an eye on updates and will open an issue if I ever see a breaking change.

👍🏻 Unreliable though should do the thing most of the time 😉 

> Go's symlink API says it supports Windows and apparently Windows has supported symlinks since Vista (which is news to me)

**Comment by gabe565** [comment]
@DariuszPorowski I appreciated your PR. The config dir logic is much cleaner now. I just released v1.3.1 which includes your PR. You will still need to run the command with elevated privileges for now, but can you give it another try?

Thank you!

**Comment by DariuszPorowski** [comment]
@gabe565 Work like a charm! :) elevated privileges it's not a big deal for me.
Symlinks can expose sec vulnerabilities in apps that aren't designed to handle them. So, I am OK with that - just one more click ;)
Uses cen enable symlinks without extra permission in local security policy `secpol.msc` but I'd rather want to keep defaults. 

More interested in: https://github.com/gabe565/gh-profile/issues/21 :D

**Comment by gabe565** [comment]
Sweet! Thanks for testing!

That makes sense with the elevated access. I'm more familiar with *nix OS's and would get nervous if a random extension needed `sudo`, but if that's less jarring in Windows then it works for now. I'd like to investigate other ways in Windows in the long term, though.

Agreed on #21 being top priority. Unfortunately, this extension is blocked by a limitation of the gh cli at the moment. When the gh cli fetches the current auth token from the OS keychain, it always uses the same service key. I'm working on adding the current user to that service key so that multiple tokens can be stored in the keychain. Once I get that merged in and they ship a release, `gh-profile` should support secure tokens without any changes!

**Comment by gabe565** [comment]
@DariuszPorowski Sweet! Thanks for testing!

That makes sense with the elevated access. I'm more familiar with *nix OS's and would get nervous if a random extension needed `sudo`, but if that's less jarring in Windows then it works for now. I'd like to investigate other ways in Windows in the long term, though.

Agreed on #21 being top priority. Unfortunately, this extension is blocked by a limitation of the gh cli at the moment. When the gh cli fetches the current auth token from the OS keychain, it always uses the same service key. I'm working on adding the current user to that service key so that multiple tokens can be stored in the keychain. Once I get that merged in and they ship a release, `gh-profile` should support secure tokens without any changes!
